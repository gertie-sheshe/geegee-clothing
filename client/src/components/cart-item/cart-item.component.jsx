import React from 'react';

import {
  CartItemContainer,
  CartItemImage,
  ItemDetailsContainer,
} from './cart-item.styles';

const CartItem = ({ item: { imageUrl, price, name, quantity } }) => (
  <CartItemContainer>
    <CartItemImage src={imageUrl} alt="item in cart" />
    <ItemDetailsContainer>
      <span>{name}</span>
      <span>
        ${price} x {quantity}
      </span>
    </ItemDetailsContainer>
  </CartItemContainer>
);

export default React.memo(CartItem);
