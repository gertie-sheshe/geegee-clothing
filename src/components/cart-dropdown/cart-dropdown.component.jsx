import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { withRouter } from 'react-router-dom';

import CartItem from '../cart-item/cart-item.component';

import { toggleCartHidden } from '../../redux/cart/cart.actions';

// Selectors
import { selectCartItems } from '../../redux/cart/cart.selectors';

import {
  CartDropdownContainer,
  CartItemContainer,
  EmptyMessage,
  CartDropdownButton,
} from './cart-dropdown.styles';

const CartDropDown = ({ cartItems, history, dispatch }) => (
  <CartDropdownContainer>
    <CartItemContainer>
      {cartItems.length ? (
        cartItems.map(cartItem => (
          <CartItem key={cartItem.id} item={cartItem} />
        ))
      ) : (
        <EmptyMessage>Your cart is empty</EmptyMessage>
      )}
    </CartItemContainer>
    <CartDropdownButton
      onClick={() => {
        history.push('/checkout');
        dispatch(toggleCartHidden());
      }}
    >
      GO TO CHECKOUT
    </CartDropdownButton>
  </CartDropdownContainer>
);

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems,
});

export default withRouter(connect(mapStateToProps, null)(CartDropDown));
