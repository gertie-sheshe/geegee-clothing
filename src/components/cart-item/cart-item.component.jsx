import React from 'react';

import './cart-item.styles.scss';

import {
  CartItemContainer,
  CartItemImage,
  ItemDetailsContainer,
} from './cart-item.styles';

const CartItem = ({ item: { imageUrl, price, name, quantity } }) => (
  <CartItemContainer>
    <CartItemImage src={imageUrl} alt="item in cart" />
    <ItemDetailsContainer>
      <span>{name}</span>
      <span>
        ${price} x {quantity}
      </span>
    </ItemDetailsContainer>
  </CartItemContainer>
);

export default CartItem;
