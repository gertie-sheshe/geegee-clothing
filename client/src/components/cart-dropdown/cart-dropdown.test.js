import React from 'react';
import { shallow } from 'enzyme';

import { CartDropDown } from './cart-dropdown.component';
import CartItem from '../cart-item/cart-item.component';

import { toggleCartHidden } from '../../redux/cart/cart.actions';
import { ItemDetailsContainer } from '../cart-item/cart-item.styles';

describe('CartDropdown Component', () => {
  let wrapper;
  let mockHistory;
  let mockDispatch;

  const mockCartItems = [{ id: 1 }, { id: 2 }, { id: 3 }];

  beforeEach(() => {
    mockHistory = {
      push: jest.fn(),
    };

    mockDispatch = jest.fn();

    const mockProps = {
      cartItems: mockCartItems,
      history: mockHistory,
      dispatch: mockDispatch,
    };

    wrapper = shallow(<CartDropDown {...mockProps} />);
  });

  it('Should render CartDropdown Component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should call history.push when button is clicked', () => {
    wrapper.find('CartDropdownButton').simulate('click');
    expect(mockHistory.push).toHaveBeenCalled();
    expect(mockDispatch).toHaveBeenCalledWith(toggleCartHidden());
  });

  it('Should render an equal number of CartItem components as the cartItems props', () => {
    expect(wrapper.find(CartItem).length).toEqual(mockCartItems.length);
  });

  it('Should render EmptyMessageContainer when cartItems is empty', () => {
    const mockProps = {
      cartItems: [],
      history: mockHistory,
      dispatch: mockDispatch,
    };

    const newWrapper = shallow(<CartDropDown {...mockProps} />);
    expect(newWrapper.exists('EmptyMessageContainer')).toBe(true);
  });
});
