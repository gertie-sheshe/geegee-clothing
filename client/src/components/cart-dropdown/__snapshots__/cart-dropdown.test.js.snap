// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDropdown Component Should render CartDropdown Component 1`] = `ShallowWrapper {}`;
